docs/specs/lcb-traits.md

```diff
@@ -0,0 +1,302 @@
+        -- Perform formatting on T
+    end handler
+
+In this case, the compiler would identify that `M.FormatAsString<T>()` has a
+name that matches a generic handler declaration, and that it conforms (all
+generic parameters are specialised and all non-generic parameters are of the
+correct types).  Subsequent calls to `FormatAsString()` from *anywhere in the
+program* would then be able to access `M.FormatAsString<T>()` if the first
+argument of `FormatAsString()` conforms to `T`.
+
+The main drawback here is that Builder is optionally-typed, but it would be
+good to ensure that it's possible to write code where all types are specified
+*and can be verified* at compile time.  This doesn't satisfy the requirement.
+
+For example, it would be convenient to write a `Log()` handler that can log
+any type that can also be formatted.  However, with the generic handlers scheme
+described above, `Log()` would have to take arguments as `any` and defer
+type-checking to runtime.
+
+## Traits
+
+Traits are collections of methods that can be used to extend the functionality
+of a type.
+
+Traits have no storage associated with them.  Storage is provided by
+implementing the trait for a specific storage type.
+
+### Defining traits
+
+    public trait type CanFormatAsString
+        method type FormatAsString(in me) as string
+    end trait
+
+A trait definition is a block containing a list zero or more `method`
+definitions or `method type` declarations.  `method type` declarations take the
+same form as existing `handler type` declarations.  `method` definitions take
+the same form as existing `handler` declarations.
+
+    public trait type CanFormatAsString
+        method FormatAsString(in me) as string
+            return "<unknown>"
+        end method
+    end trait
+
+`method` definitions allow a trait's definition to include default
+implementations of the trait's methods.  Default implementations can call other
+methods of the trait or perform any operations valid for the generic `Self`
+type.
+
+Two additional symbols are defined within a trait definition:
+
+1. `MyType` is always equivalent to the type of the value that the method is
+   being called for.
+2. `me` must always be the first parameter a `method` definition or
+   `method type` declaration; it is equivalent to writing `me as MyType`.  The
+   `me` parameter must always be `in` or `inout`.
+
+### Implementing traits
+
+    public implementation of CanFormatAsString for T
+        method FormatAsString(in me) as string
+            variable tFormatted as string
+            -- Format me parameter
+            return tFormatted
+        end method
+    end implementation
+
+The `me` and `MyType` symbols are also defined inside implementation blocks.
+
+In order for an `implementation of U for T` to be valid, the following criteria
+must be fulfilled:
+
+1. Either `U` or `T` must be local, i.e. defined in the current module.  This
```

Comment:
Cycles aren't allowed, no - the reason I bring this up is that it will influence how versioning needs to work as this seems like a potentially reasonable thing to try and accommodate.

Traits combined with the strict module namespace essentially de-centralize policy, but the rule about placement of traits, types and implementations of types means that it would be difficult for a type producer to adopt a trait natively that had previously been implemented externally (which they may want to do if they can implement it more efficiently - I'm assuming here that types will be opaque outside of the module which produces them, and thus only the defining module would have access to its internals).

If you imagine a module being composed of separate parts (source-files) and versions can be specified per-part (rather than per-definition), then the compiler would compile the module parts in version order - first producing the version 1 interface, then the version 2 interface etc. The idea here being that when it compiles version 2 it can use other modules which only depend on version 1.

Perhaps there's another way to achieve the above - or maybe I'm overthinking things...